### ISSUE TITLE
Question: Stability and plans for this crate

### ISSUE THREAD

**Issue by weigangd** [minghuaw/fe2o3-amqp#105]
Hello @minghuaw,

first of all thank you for creating this crate. I think it useful and works really nice. Therefore I probably want to use it more often (mostly the client side). 
Would you say that this crate reached a certain point of stability or do you think that there will be a lot of breaking changes in the future? What would be your plans for the future of this crate?
Thank you really much :)

**Comment by minghuaw** [comment]
Thank you for your interest in this crate :). I will try to answer your concerns below.

## Stability

Because 99% of the core protocol is already implemented, I don't think there will be much (if any) breaking changes in the near future, and it's probably going to stay in the `"0.6.x"` version for a while.

## Plan

As for the plan for the future of this crate, I am going to keep actively developing and maintaining this crate as well as other AMQP related crates in the foreseeable future. Though I don't have a concrete timeline planned, the followings are likely going to be what I will work on

1. Adding more unit testing (though I don't picture any breaking bug fixes) 
2. Implement AMQP client for the unofficial `azure-sdk-for-rust/messaging_servicebus` SDK and potentially another SDK for `event_hubs` (this will likely take some time though, as I am not affiliated with microsoft, I don't have access to how azure encode some of its custom message properties. Plus, azure uses some of the extension specs that are still working drafts, which leads to the management and CBS extensions discussed below)
3. Implement extension specs and working drafts
    - [x] WebSocket binding. Though it's only `"0.1.x"`, the spec is only asking to treat WebSocket as if it is a stream of bytes, and everything's working fine. So it is stable enough and will probably remain as it is for while. (https://github.com/minghuaw/fe2o3-amqp/tree/main/fe2o3-amqp-ws)
    - [x] Filters. I will add new filters to this crate if there are new additions in the specs/working draft. (https://github.com/minghuaw/fe2o3-amqp/tree/main/fe2o3-amqp-ext)
    - [ ] Management extension working draft. The working draft is very much work-in-progress itself and isn't clear on the specifics, so treat it as **unstable**. (https://github.com/minghuaw/fe2o3-amqp/tree/management/fe2o3-amqp-management)
    - [ ] CBS extension working draft. This kind of relies on the management extension and is in a similar state as the management extension.

Bug reports and feature requests are always welcome. If you need a specific SDK that uses AMQP (eg. azure servicebus/eventhubs) and couldn't find one, I will be happy to implement that.

**Comment by minghuaw** [comment]
Another note is that supporting more async runtime may introduce breaking change (likely just a new feature in your Cargo.toml), but i don't think this will be  done in the near future do the following reasons. 

1. I haven't found a way to achieve this without duplicating . 
2. There is a bug in tokio_util codec that won't be fixed soon, and i as of now don't know whether its async-std counterpart has the same bug
3. There were some news on tokio with wasm recently, which kind of removes one area where async-std might be the better option

**Comment by weigangd** [comment]
Stability-wise it sounds great, that most probably only few (if any) breaking changes are to be expected. Also the plan looks well thought out. Great work!
Thank you for the quick and detailed reply! :)

**Comment by minghuaw** [comment]
@weigangd While trying to implement an AMQP client for servicebus, I found that one thing that I intentionally omitted (#112) was actually required for servicebus sdk. There is an experimental implementation (in the [`"batch"` branch](https://github.com/minghuaw/fe2o3-amqp/blob/da830747591bad7d0e411a2e39decedc978f8307/fe2o3-amqp-types/src/messaging/message/body.rs#L16)) which essentially make message body like the following (the two new additions are `DataBatch` and `SequenceBatch` variants).

```rust
pub enum Body {
    Data(Data(_)),
    Sequence(AmqpSequence(_)),
    Value(AmqpValue(_)),
    // `TransparentVec` is a thin wrapper over `Vec` but not encoded as a list
    DataBatch(TransparentVec<Data>), 
    // `TransparentVec` is a thin wrapper over `Vec` but not encoded as a list
    SequenceBatch(TransparentVec<AmqpSequence>),
    Empty,
}
```

I chose to make two new variants instead of changing the original `Data` and `Sequence` variants because the batch operations are not that common, and I don't want to make the users always taking element from a `Vec` of a single element. Hopefully this justified my choice.

Here comes the problem, this technically should be considered a breaking change, but unless the user was directly matching each variant of `Body`, everything should remain the same. So it might mean a `"0.7.0" release today or tomorrow. 

I am just trying to see what's your view on this potential version bump? Hopefully it wouldn't affect your usage much.

**Comment by minghuaw** [comment]
Another option is to gate the `***Batch` variants behind a new feature flag that is disabled by default, then it could be a smooth "0.6.10" upgrade unless the feature is enabled. What do you think?

**Comment by minghuaw** [comment]
@weigangd I might have to make PR #113 a breaking change (ie. version "0.7.0"). You wouldn't feel any breaking change if there is no direct matching against message `Body`. I have included some discussions in the issue #112. 

Sorry that I broke my word on "0.6.x staying here for a while", but I would love to hear what you think about it. I will probably wait a couple days before I merge the PR.

**Comment by weigangd** [comment]
@minghuaw First of all thank you for reaching out and also sorry for not seeing this earlier!

My question was not about ensuring that there will be no breaking changes in the near future, it was just about knowing what your plans are going to be :). I am a big fan of [Semantic Versioning](https://semver.org) but even in their specification they say
> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.
so I just wanted to see how much is probably stable or if I should expect **everything** to change soon (which is clearly not the case here :) )
Regarding the change and a version bump: I think it is a reasonable choice and I would totally welcome it. I would also be against using a feature-gate for the same reason you mentioned in #112. Regarding using `#[non-exhaustive]` I would ask myself if there is a high probability if new variants are introduced in the future. If that is not the case then I don't know if this is necessary.

Breaking changes happen, so don't hesitate if there are good reasons for them :)

**Comment by minghuaw** [comment]
@weigangd Thanks for your feedback. So I ended up changing message body to a trait and then restrict the implementation to a couple of types which can then be converted into/from for other types. More details can be found in `fe2o3-amqp-types` changelog (currently in "0.7.0-dev" branch, https://github.com/minghuaw/fe2o3-amqp/blob/0.7.0-dev/fe2o3-amqp-types/Changelog.md) or PR #116 . And an alpha preview has been published to crates.io.

Looking back now, the reason why I was hesitant was not really changing the version number but the old approach of representing message body as a single `enum` left me in an awkward situation. More specifically, a single `AmqpSequence` or `Data` body section is probably much more frequent than multiple `AmqpSequnce`/`Data` sections, and I was really debating whether to separate a single `AmqpSequence`/`Data` away from multiple of them by having separate variants for single body section and multiple body section. 

Now, after I ended up with a changing message body to a generic that is only serializable/deserializable when it meets the trait bound, this no longer feels like a problem to me. It is more expressive in terms of the exact body section type to use.

Thanks again. I think there's still some room for more testing cases before making it officially "0.7.0". Hopefully I will get this worked out soon.
